examples/project_llm_fine_tune/project_llm_fine_fune/assets.py

```diff
@@ -0,0 +1,395 @@
+    validation_file,
+) -> str:
+    with openai.get_client(context) as client:
+        with open(validation_file, "rb") as file_fd:
+            response = client.files.create(file=file_fd, purpose="fine-tune")
+    return response.id
+
+
+@dg.asset(
+    kinds={"OpenAI"},
+    description="Exeute model fine-tuning job",
+    group_name="fine_tuning",
+)
+def fine_tuned_model(
+    context: dg.AssetExecutionContext,
+    openai: OpenAIResource,
+    upload_training_file,
+    upload_validation_file,
+) -> str:
+    with openai.get_client(context) as client:
+        create_job_resp = client.fine_tuning.jobs.create(
+            training_file=upload_training_file,
+            validation_file=upload_validation_file,
+            model=constants.MODEL_NAME,
+            suffix="goodreads",
+        )
+
+    create_job_id = create_job_resp.to_dict()["id"]
+    context.log.info(f"Fine tuning job: {create_job_id}")
+
+    while True:
+        status_job_resp = client.fine_tuning.jobs.retrieve(create_job_id)
+        status = status_job_resp.to_dict()["status"]
+        context.log.info(f"Job {create_job_id}: {status}")
+        if status in ["succeeded", "cancelled", "failed"]:
+            break
+        time.sleep(30)
+
+    fine_tuned_model_name = status_job_resp.to_dict()["fine_tuned_model"]
+    context.add_output_metadata({"model_name": fine_tuned_model_name})
+    return fine_tuned_model_name
+
+
+def model_question(client: OpenAI, model: str, prompt: str, question: str) -> bool:
+    completion = client.chat.completions.create(
+        model=model,
+        messages=[
+            {
+                "role": "system",
+                "content": dedent(prompt),
+            },
+            {"role": "user", "content": dedent(question)},
+        ],
+        response_format={
+            "type": "json_schema",
+            "json_schema": {
+                "name": "is_manga",
+                "schema": {
+                    "type": "object",
+                    "properties": {"is_manga": {"type": "boolean"}},
+                    "required": ["is_manga"],
+                    "additionalProperties": False,
+                },
+                "strict": True,
+            },
+        },
+    )
+    content = completion.choices[0].message.content
+    return json.loads(content)["is_manga"]
+
+
+@dg.asset_check(
+    asset=fine_tuned_model,
+    description="Compare fine-tuned model against base model accuracy",
+)
+def fine_tuned_model_accuracy(
+    context: dg.AssetCheckExecutionContext,
+    duckdb_resource: dg_duckdb.DuckDBResource,
+    openai: OpenAIResource,
+    fine_tuned_model,
+) -> Iterable[dg.AssetCheckResult]:
+    models = Counter()
+    prompt = """
+        You will be provided with the title, author and description of a book.
+        Determine if the title is or is not manga.
+    """
+    query = f"""
+        select * from enriched_graphic_novels
+    """
+    with duckdb_resource.get_connection() as conn:
```

Comment:
One thing I couldn't figure out with an asset check, if it's possible to use the output of an asset that is not part of the check. Having to reread with DuckDB but if you could just take in `enriched_graphic_novels` it would save some lines.